services/agents/handler.go

```diff
@@ -0,0 +1,332 @@
+	defaultQueryActionTimeout = durationpb.New(15 * time.Second) // should be less than checks.resultTimeout
+	defaultPtActionTimeout    = durationpb.New(30 * time.Second) // Percona-toolkit action timeout
+
+	mSentDesc = prom.NewDesc(
+		prom.BuildFQName(prometheusNamespace, prometheusSubsystem, "messages_sent_total"),
+		"A total number of messages sent to pmm-agent.",
+		[]string{"agent_id"},
+		nil,
+	)
+	mRecvDesc = prom.NewDesc(
+		prom.BuildFQName(prometheusNamespace, prometheusSubsystem, "messages_received_total"),
+		"A total number of messages received from pmm-agent.",
+		[]string{"agent_id"},
+		nil,
+	)
+	mResponsesDesc = prom.NewDesc(
+		prom.BuildFQName(prometheusNamespace, prometheusSubsystem, "messages_response_queue_length"),
+		"The current length of the response queue.",
+		[]string{"agent_id"},
+		nil,
+	)
+	mRequestsDesc = prom.NewDesc(
+		prom.BuildFQName(prometheusNamespace, prometheusSubsystem, "messages_request_queue_length"),
+		"The current length of the request queue.",
+		[]string{"agent_id"},
+		nil,
+	)
+)
+
+// Handler handles agent requests.
+type Handler struct {
+	db          *reform.DB
+	r           *Registry
+	qanClient   qanClient
+	jobsService *JobsService
+
+	mConnects    prom.Counter
+	mDisconnects *prom.CounterVec
+	mRoundTrip   prom.Summary
+	mClockDrift  prom.Summary
+}
+
+// NewHandler creates new agents handler.
+func NewHandler(db *reform.DB, qanClient qanClient, registry *Registry, service *JobsService) *Handler {
+	h := &Handler{
+		db:          db,
+		qanClient:   qanClient,
+		r:           registry,
+		jobsService: service,
+
+		mConnects: prom.NewCounter(prom.CounterOpts{
+			Namespace: prometheusNamespace,
+			Subsystem: prometheusSubsystem,
+			Name:      "connects_total",
+			Help:      "A total number of pmm-agent connects.",
+		}),
+		mDisconnects: prom.NewCounterVec(prom.CounterOpts{
+			Namespace: prometheusNamespace,
+			Subsystem: prometheusSubsystem,
+			Name:      "disconnects_total",
+			Help:      "A total number of pmm-agent disconnects.",
+		}, []string{"reason"}),
+		mRoundTrip: prom.NewSummary(prom.SummaryOpts{
+			Namespace:  prometheusNamespace,
+			Subsystem:  prometheusSubsystem,
+			Name:       "round_trip_seconds",
+			Help:       "Round-trip time.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		}),
+		mClockDrift: prom.NewSummary(prom.SummaryOpts{
+			Namespace:  prometheusNamespace,
+			Subsystem:  prometheusSubsystem,
+			Name:       "clock_drift_seconds",
+			Help:       "Clock drift.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		}),
+	}
+
+	// initialize metrics with labels
+	h.mDisconnects.WithLabelValues("unknown")
+	return h
+
+}
+
+// Run takes over pmm-agent gRPC stream and runs it until completion.
+func (h *Handler) Run(stream agentpb.Agent_ConnectServer) error {
+	h.mConnects.Inc()
+	disconnectReason := "unknown"
+	defer func() {
+		h.mDisconnects.WithLabelValues(disconnectReason).Inc()
```

Comment:
and include this into `unregister` in `registry`, disconnectReason might be one more arg.

Reply:
what about 
```
h.mRoundTrip.Observe(roundtrip.Seconds())
	h.mClockDrift.Observe(clockDrift.Seconds())
```
or should move ping method to registry?

Reply:
Yeah, I was thinking about moving `ping` into registry, since it's related to connection and registry.